Project blog: loop flattening in LLVM for Proposal #222

hj424 opened the pull request.

hj424 committed:
add blog post for final project

hj424 committed:
Update index.md

hj424 committed:
Update index.md

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+
+## Background
+1. LLVM Compiler Infrastructure<br/>
+This open-source project was started by the researchers from UIUC in 2000. It is a collection of the modular and reusable compiler. The goal of this project is to provide a modern, SSA-based compilation strategy capable of supporting both static and dynamic compilation of arbitrary programming languages. With the hard work from a diverse community, the LLVM compiler has become a powerful and stab ...
```
> It is a collection of the modular and reusable compiler.

Add "components"?

sampsyo replied:
I'd just link to the LLVM homepage right here instead of listing references at the end—that will be more helpful for your readers.

sampsyo replied:
This sentence:

> The goal of this project is to provide a modern, SSA-based compilation strategy capable of supporting both static and dynamic compilation of arbitrary programming languages.

and some other phrases seem to be taken from the LLVM homepage. No need to copy this text. Just link to the website where people can read it for themselves and give your own short overview her.

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+[[extra.authors]]
+name = "Hanchen Jin"
++++
+
```
Before jumping into the background, it would be great to give a short overview of what this blog post will be about.

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+This open-source project was started by the researchers from UIUC in 2000. It is a collection of the modular and reusable compiler. The goal of this project is to provide a modern, SSA-based compilation strategy capable of supporting both static and dynamic compilation of arbitrary programming languages. With the hard work from a diverse community, the LLVM compiler has become a powerful and stab ...
+
+2. Loop Optimizations<br/>
+As a programmer, we always want to finish running our programs as fast as possible. However, with the end of Moore’s law around 2000, the computing power for a single CPU is limited by its physical features. To further improve the performance, the multi-core system with parallel computing has become a popular solution. But it requires the programmer to manually optimize the code to achieve higher ...
```
can -> to, in "want the compiler itself can figure"
heated -> hot

sampsyo replied:
I'm a little confused about the discussion here of multicore processors (and the similar discussion two paragraphs down). This optimization doesn't use multiple cores, does it? All the code here is single-threaded. So I'm not sure Moore's law, etc., is really relevant as a motivation.

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+2. Loop Optimizations<br/>
+As a programmer, we always want to finish running our programs as fast as possible. However, with the end of Moore’s law around 2000, the computing power for a single CPU is limited by its physical features. To further improve the performance, the multi-core system with parallel computing has become a popular solution. But it requires the programmer to manually optimize the code to achieve higher ...
+
+As we know, loops are usually the heavy part during execution, which consumes most of the runtime. Thus, many strategies have been applied to speed up the execution time of the loops. For example, Loop Invariant Code Motion (LICM) removes the loop invariant instructions out of the loop to avoid perform useless instructions. For more details about all available loop optimizations, please check out ...
```
"all available loop optimizations": available in LLVM specifically, right?

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+As we know, loops are usually the heavy part during execution, which consumes most of the runtime. Thus, many strategies have been applied to speed up the execution time of the loops. For example, Loop Invariant Code Motion (LICM) removes the loop invariant instructions out of the loop to avoid perform useless instructions. For more details about all available loop optimizations, please check out ...
+
+## Motivation
+As we mentioned before, with the multi-core compute architecture, we would like to leverage the parallelism of the code to get better performance. As for some loops, there is no data dependency around multiple iterations, which is a good fit for loop unrolling optimization. Specifically, multiple loop iterations can be executed independently at the same time. Therefore, in this section, we will i ...
```
I guess what I'm getting at here is that it's true that "multiple loop iterations can be executed independently at the same time," but if we just unroll the loop, we're relying on a superscalar processor to do that—not a multicore processor.

In other words, unrolling enhances ILP, not TLP.

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
````diff
@@ -0,0 +1,199 @@
+
+```javascript
+for (int i = 0; i < n*m; i+= 2) { // flattened loop
+  Stmt(f(i))
````
It's not entirely clear what's going on here—`f` used to be a two-argument function, and now it's a one-argument function. Did something have to change inside `f`?

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
````diff
@@ -0,0 +1,199 @@
+}
+```
+
+After flattening the loop, the loop bound becomes larger so that we can apply larger unrolling parameters to further explore the parallelism. Also, this technique is necessary for some optimizations in High-Level Synthesis tools. For example, suppose we have a nested for loop with variable loop bound, the tool cannot automatically apply the pipeline optimization because it has no knowledge about  ...
````
In "variable loop bound, the tool cannot" replace the comma with a colon or semicolon or start a new sentence.

sampsyo replied:
"importance of the loop flatten technique" -> loop flattening

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+
+After flattening the loop, the loop bound becomes larger so that we can apply larger unrolling parameters to further explore the parallelism. Also, this technique is necessary for some optimizations in High-Level Synthesis tools. For example, suppose we have a nested for loop with variable loop bound, the tool cannot automatically apply the pipeline optimization because it has no knowledge about  ...
+
+
```
After reading this section, I'm a little confused about why the post also describes unrolling and unroll-and-jam in addition to flattening. Seems like flattening is the focus here? Is there some additional context to explain why the other optimizations get demonstrated too?

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+The idea is quite straightforward but the implementation is not that easy. To implement this pass, we should perform the following modifications.
+
+1. Update the loop bound and increment <br/>
+As the first step, we should update the loop bound for the outer loop. Suppose the loop bound for the inner loop and outer loop are m and n with the same increment, we can change the loop bound for the outer loop to be `m*n` and keep the same increment.
```
"increment, we": Again, use : or ; or start a new sentence.

sampsyo replied:
Be consistent about whether backticks are used for variables. For example, if `m*n` is going to be in code backticks, `m` and `n` should be too.

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+As the first step, we should update the loop bound for the outer loop. Suppose the loop bound for the inner loop and outer loop are m and n with the same increment, we can change the loop bound for the outer loop to be `m*n` and keep the same increment.
+
+2. Extract the statements inside the inner loop to the outer loop<br/>
+In this step, we extract all the statements inside the inner loop and put them into the outer loop. At the same time, we should detect the usage of the inner loop variable and then remove it with the correct one after flattening the loop. For example, suppose we have a piece of code like this inside the inner loop: `i*n+j`, whereas `i` and `j` are the loop variables, `n` is the outer loop bound,  ...
```
remove -> replace

sampsyo replied:
It makes sense that you'd replace `i*n+j` with `i`, as an example. But this description leaves it somewhat unclear what the general rule is. Is this *only* going to work with `i*n+j`, or is it possible to use `i` and `j` independently also? A little clarity beyond the single example to shed light on the general rule would be helpful.

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+
+Finally, we should check the equivalence of the code, i.e., the flattened loop should perform the same execution as the original nested loop. This can be verified by testing the pass with benchmarks. 
+
+During the implementation, I got stuck in the second step, i.e., extract the statements inside the inner loop to the outer loop. The generated code performs weird when extracts the loop variant expressions from the inner loop to the outer loop. 
```
"when extracts" -> when extracting

sampsyo replied:
What exactly goes wrong? Some specificity here about what "weird" thing happened would be useful.

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+
+## Evaluation
+1. Correctness <br/>
+As for the correctness of the pass, we should focus on two points: the functionality of the optimized code and the correctness of the unrolled loop. These two points are being checked by running the simple benchmarks with the clang profiling tools. For more details, please check out [this link](https://clang.llvm.org/docs/SourceBasedCodeCoverage.html).
```
"the functionality of the optimized code and the correctness of the unrolled loop" sound like the same thing? Maybe this is supposed to be the correctness and the performance?

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
````diff
@@ -0,0 +1,199 @@
+
+To elaborate the evaluation, let me use the following source code as an example: 
+
+```javascript
````
Maybe use C for better syntax highlighting?

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+The first line is the sample result I printed out to verify the correctness of the optimized code. And then the following code is the optimized one with the updated inner loop bound. As you can see from line 11, the outer loop was being executed only 1 time. The number is 2 here means the second time, it checks the loop bound and exits the outer loop. And accordingly, the inner loop is being exec ...
+
+2. Performance gain<br/>
+As we mentioned before, the loop flattening pass itself does not offer any benefits for performance. But this technique helps to better explore the parallelism when combining with the loop unrolling technique. Specifically, we can apply larger unroll parameters to further explore the parallelism across the loop iterations. Also, this technique is used to support other optimizing pragmas (like loo ...
```
It would have been interesting, FWIW, to have measured the impact on these downstream optimizations.

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+
+
+## Conclusion
+In summary, loop flattening is a useful optimizing technique for supporting loop unrolling and other HLS related optimizations. LLVM is an awesome infrastructure that provides us powerful APIs for creating our own pass. Finally, I would like to mention that the loop flattening pass has just been merged into the official LLVM github repository (LLVM v12.0.0). Here is the [source code](https://llvm ...
```
I would say "*a* loop flattening pass" instead of "*the* loop flattening pass" to avoid anyone thinking that this was _your_ pass that got merged. :smiley:

sampsyo replied:
Capitalize GitHub.

sampsyo commented on content/blog/2020-12-20-loop-flatten/index.md:
```diff
@@ -0,0 +1,199 @@
+[2] LLVM user reference manual: https://llvm.org/docs/LangRef.html<br/>
+[3] LLVM developer reference manual: http://llvm.org/doxygen/<br/>
+[4] LLVM official github repository: https://github.com/llvm/llvm-project<br/>
+[5] Loop flatten techniques applied in software pipelining (converting loop level parallelism into instruction parallelism): https://people.ece.uw.edu/hauck/publications/LoopFlattening.pdf<br/>
```
I generally think it's probably better to link to stuff in context, when you're writing for the Web. Unlike when you're writing a PDF, it's just as easy to do that and more convenient to read than having references at the end.

sampsyo reviewed (changes requested):
Interesting experience report! In general, the implementation seems solid if limited, but I would have liked to see a more complete evaluation. Showing that the pass at least does not break some real-world programs would have been useful, for instance. Here are some suggestions for the writing.

hj424 committed:
Update index.md

1. Fix the writing issue.
2. Add the checking module in the pass to make sure this pass works for real-world programs, i.e., it will not destroy the code with non-perfect nested loops.
3. Verify the pass works for some benchmarks from embed-iot

hj424 commented:
Hi Adrian,

Thanks for the suggestions! I have updated my blog in the latest post. Here are the keynotes of my update:
1. Fix the writing issue.
2. Add the checking module in the LLVM pass to make sure this pass works for real-world programs, i.e., it will not destroy the code with non-perfect nested loops.
3. Verify that the pass works for some benchmarks from embed-iot

Thanks,
Hanchen

sampsyo commented:
Sounds good! But there are outstanding comments about nitpicky syntax things. Please request a review once those are fixed.

hj424 committed:
Fix syntax issue and clarify the confusing part

1. Fix the syntax issue.
2. Clarify the confusing part.

hj424 commented:
Hi Adrian,

Thanks for the notification. In the latest commit, I have fixed the syntax issue and clarified the confusing part. Please take a look.

Thanks,
Hanchen

sampsyo reviewed (approved):
Looks good; thanks!
